runtime/lua/vim/shared.lua

```diff
@@ -1151,34 +1154,70 @@ end
+    end
+  end
+
+  -- Set other scopes (so that they have more precedence over `context.o`)
+  for scope, scope_cache in pairs(cache) do
+    for name, context_value in pairs(context[scope] or {}) do
+      -- Do not override already set (from `context.o`) cache
+      scope_cache[name] = scope_cache[name] or vim[scope][name]
+      vim[scope][name] = context_value
+    end
+  end
+
+  return cache
+end
 
 --- Executes function `f` with the given context specification.
 ---
 --- @param context vim.context.mods
 function vim._with(context, f)
-  vim.validate('context', context, 'table')
-  vim.validate('f', f, 'function')
-
-  vim.validate('context.buf', context.buf, 'number', true)
-  vim.validate('context.emsg_silent', context.emsg_silent, 'boolean', true)
-  vim.validate('context.hide', context.hide, 'boolean', true)
-  vim.validate('context.horizontal', context.horizontal, 'boolean', true)
-  vim.validate('context.keepalt', context.keepalt, 'boolean', true)
-  vim.validate('context.keepjumps', context.keepjumps, 'boolean', true)
-  vim.validate('context.keepmarks', context.keepmarks, 'boolean', true)
-  vim.validate('context.keeppatterns', context.keeppatterns, 'boolean', true)
-  vim.validate('context.lockmarks', context.lockmarks, 'boolean', true)
-  vim.validate('context.noautocmd', context.noautocmd, 'boolean', true)
-  vim.validate('context.options', context.options, 'table', true)
-  vim.validate('context.sandbox', context.sandbox, 'boolean', true)
-  vim.validate('context.silent', context.silent, 'boolean', true)
-  vim.validate('context.unsilent', context.unsilent, 'boolean', true)
-  vim.validate('context.win', context.win, 'number', true)
+  vim.validate({
+    context = { context, 'table' },
+    f = { f, 'function' },
+  })
+
+  vim.validate({
+    ['context.bo'] = { context.bo, 'table', true },
+    ['context.buf'] = { context.buf, 'number', true },
+    ['context.emsg_silent'] = { context.emsg_silent, 'boolean', true },
+    ['context.env'] = { context.env, 'boolean', true },
+    ['context.go'] = { context.go, 'table', true },
+    ['context.hide'] = { context.hide, 'boolean', true },
+    ['context.horizontal'] = { context.horizontal, 'boolean', true },
+    ['context.keepalt'] = { context.keepalt, 'boolean', true },
+    ['context.keepjumps'] = { context.keepjumps, 'boolean', true },
+    ['context.keepmarks'] = { context.keepmarks, 'boolean', true },
+    ['context.keeppatterns'] = { context.keeppatterns, 'boolean', true },
+    ['context.lockmarks'] = { context.lockmarks, 'boolean', true },
+    ['context.noautocmd'] = { context.noautocmd, 'boolean', true },
+    ['context.o'] = { context.o, 'table', true },
+    ['context.sandbox'] = { context.sandbox, 'boolean', true },
+    ['context.silent'] = { context.silent, 'boolean', true },
+    ['context.unsilent'] = { context.unsilent, 'boolean', true },
+    ['context.win'] = { context.win, 'number', true },
+    ['context.wo'] = { context.wo, 'table', true },
+  })
```

Comment:
Out of curiosity, is there a reason to change this?

The single-line `vim.validate` calls are basically free. I don't think changing this is a readability win either way (there's a small argument to be made for eliminating the repeated `vim.validate`, but that's it). My main concern is about establishing patterns. Future contributors will copy-paste what they find elsewhere, and I think we should try to use the fast version of `vim.validate` wherever possible for that reason.

Reply:
My bad, it was more of an assumption that single call to `vim.validate()` is more performant than multiple. However, I did not account that those are the newly added style of vararg calls instead of creating a table every single time.

I've profiled two versions (single `vim.validate()` with table argument vs multiple with vararg) and indeed vararg version seems to be faster: around 5e-07 seconds for single `vim.validate()` and around 8e-08 seconds for vararg (time for all current 19 calls).

I'll change back to multiple vararg calls.